### DISCUSSION TITLE
.p12 certificates created with versions 3.0.11, 3.0.13 and 3.1.5 cannot be imported to Chromium or Brave on Linux

### DISCUSSION THREAD

**Discussion by nikmit** [openssl/openssl#23601] [Q&A] [answered]
The exact same set of commands used to generate a p12 certificate on a host with OpenSSL 3.0.11, produces a different result compared to what comes out on a host with OpenSSL 1.1.1f

The .p12 from the newer versions can be imported to Firefox and Edge but on Chromium or Brave errors with 'Invalid or corrupt file'
The .p12 from 1.1.1f (Linux Mint) works in all browsers.

All tested browsers are running on a Linux host with openssl 1.1.1f so this may be an issue with backwards compatibility, I have not been able to test behavour with a browser on a system with openssl 3.X

The failure was seen with openssl 3.0.11 (Debian) and with locally compiled versions 3.0.11 / 3.0.13 / 3.1.5
Compiled on a Debian 12 VM with:
```
./config --prefix=/usr/local/ssl --openssldir=/usr/local/ssl shared zlib
make
make test
make install
openssl version
```

I tried many combinations of cert parameters while debugging to no effect. I also saw an issue here suggesting there may be different behaviour when the -in to pkcs12 is a PEM file as opposed to separate -in and -inkey parameters, that didn't make a difference either.

```
openssl genpkey -algorithm RSA -out ca.key
openssl req -new -x509 -days 3650 -key ca.key -out ca.crt
openssl genpkey -algorithm RSA -out server.key
openssl req -new -key server.key -out server.csr
openssl x509 -req -days 365 -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt
openssl pkcs12 -export -out server.p12 -inkey server.key -in server.crt -CAfile ca.crt

openssl version -a
OpenSSL 3.0.11 19 Sep 2023 (Library: OpenSSL 3.0.11 19 Sep 2023)
built on: Mon Oct 23 17:52:22 2023 UTC
platform: debian-amd64
options:  bn(64,64)
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -fzero-call-used-regs=used-gpr -DOPENSSL_TLS_SECURITY_LEVEL=2 -Wa,--noexecstack -g -O2 -ffile-prefix-map=/build/reproducible-path/openssl-3.0.11=. -fstack-protector-strong -Wformat -Werror=format-security -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG -Wdate-time -D_FORTIFY_SOURCE=2
OPENSSLDIR: "/usr/lib/ssl"
ENGINESDIR: "/usr/lib/x86_64-linux-gnu/engines-3"
MODULESDIR: "/usr/lib/x86_64-linux-gnu/ossl-modules"
Seeding source: os-specific
CPUINFO: OPENSSL_ia32cap=0xdefa220b478bffff:0x842529
```

**Comment by mattcaswell** [comment, marked as answer]
The default encryption algorithm, key derivation function and digest were changed in 3.0 compared to 1.1.1 to use more modern and secure algorithms:

https://github.com/openssl/openssl/blob/d597b46f9bdb533761e36fcf1d96ce83f3f6f04d/CHANGES.md?plain=1#L1845-L1848

Please try adding the "-legacy" option to your pkcs12 command. See the man page here:

https://www.openssl.org/docs/man3.0/man1/openssl-pkcs12.html

**Comment by nikmit** [comment]
Thank you!
That totally worked and concludes 2 full days of digging and testing 😅
